analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaIntegration.java

```diff
@@ -60,25 +89,119 @@
+      trackViewedProductCategory(track);
+    } else if (ADDED_PRODUCT.equalsIgnoreCase(event)) {
+      trackAddedProduct(track);
+      trackAddedProductCategory(track);
+    } else if (COMPLETED_ORDER.equalsIgnoreCase(event)) {
+      trackCompletedOrder(track);
+    }
+  }
+
+  void trackViewedProductCategory(TrackPayload track) {
+    String category = track.properties().category();
+
+    if (isNullOrEmpty(category)) {
+      category = NONE;
+    }
+
+    Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+    Set<String> categoriesViewed;
+    if (userAttributes.containsKey(CATEGORIES_VIEWED)) {
+      categoriesViewed = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+        protected boolean removeEldestEntry(Map.Entry<String, Boolean> eldest) {
+          return size() > MAX_CATEGORIES_VIEWED_ENTRIES;
+        }
+      });
+      String serializedCategories = userAttributes.get(CATEGORIES_VIEWED);
+      String[] categories = serializedCategories.split(",");
+      categoriesViewed.addAll(Arrays.asList(categories));
+    } else {
+      categoriesViewed = new HashSet<>();
+    }
+    categoriesViewed.add(category);
+
+    userAttributes.put(CATEGORIES_VIEWED, TextUtils.join(",", categoriesViewed));
+    userAttributes.put(LAST_VIEWED_CATEGORY, category);
+    KahunaAnalytics.setUserAttributes(userAttributes);
+  }
+
+  void trackViewedProduct(TrackPayload track) {
+    String name = track.properties().name();
+    if (!isNullOrEmpty(name)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_VIEWED_NAME, name);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  void trackAddedProduct(TrackPayload track) {
+    String name = track.properties().name();
+    if (!isNullOrEmpty(name)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_NAME, name);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  void trackAddedProductCategory(TrackPayload track) {
+    String category = track.properties().category();
+    if (!isNullOrEmpty(category)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_CATEGORY, category);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  void trackCompletedOrder(TrackPayload track) {
+    double discount = track.properties().discount();
+    Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+    userAttributes.put(LAST_PURCHASE_DISCOUNT, String.valueOf(discount));
+    KahunaAnalytics.setUserAttributes(userAttributes);
+  }
+
+  @Override public void screen(ScreenPayload screen) {
+    super.screen(screen);
+
+    if (trackAllPages) {
+      KahunaAnalytics.trackEvent(String.format(VIEWED_EVENT_FORMAT, screen.event()));
```

Comment:
Screen names are optional, for a sports category, this would become `Viewed Sports Page`, which is better than swallowing the event.

Alternatively Amplitude (see `screen` method in `AmplitudeIntegration.java`) has a setting to track named pages and to track categorized pages only. In practice however, we've found it's not useful to users.